build(docs-infra): use case-insensitive encoding for content files

petebacondarwin opened the pull request:
~~Blocked on #42509.~~

To avoid having content files that have the same file path on case-insensitive
file-systems, we now encode the paths to remove uppercase characters.

petebacondarwin committed:
build(docs-infra): use case-insensitive encoding for content files

To avoid having content files that have the same file path on case-insensitive
file-systems, we now encode the paths to remove uppercase characters.

mary-poppins commented:
You can preview 83d0b87 at https://pr42414-83d0b87.ngbuilds.io/.

mary-poppins commented:
You can preview 1385b06 at https://pr42414-1385b06.ngbuilds.io/.

gkalpak reviewed (commented):
Should we revert the changes that are no longer needed from #41788 and #41842?

petebacondarwin committed:
build(docs-infra): remove `disambiguator` doc-type

Now that we disambiguate files by encoding the outputPath, there is
no need for the `disambiguator` doc-type.

petebacondarwin commented:
> Should we revert the changes that are no longer needed from #41788 and #41842?

Good call @gkalpak

- I think the second PR contains just this commit that we should revert ce2e3b54f9d8d3f438d7865ee6ec38f4b497754b.
- the first PR effectively gets reverted as part of the natural changes of this PR - I've added an additional commit to remove the `disambiguator` doc-type.

petebacondarwin committed:
Revert "fix(docs-infra): do not redirect disambiguated URLs"

This reverts commit 61ad68a586135a4d7b2b920fda0cb4238c6422fe, since there
are no longer any disambiguated paths that need special handling.

petebacondarwin commented:
@gkalpak - are there going to be any issues with SW updates, given that we are changing the names of the JSON files that need to be requested?

mary-poppins commented:
You can preview 151d0eb at https://pr42414-151d0eb.ngbuilds.io/.

gkalpak reviewed (changes requested):
> Are there going to be any issues with SW updates, given that we are changing the names of the JSON files that need to be requested?

Oh, yes :grin:

If someone visits the API list page (or an API URL directly) while having an old SW installed, they will get the old `main.js` bundle, which contains a `DocumentService` that does not know how to translate case-sensitive URLs to the case-insensitive, underscore-encoded ones. So, it will request `FormBuilder.json` instead of `f_ormb_uilder.json` and fail (unless the user has visited the specific API page before and the SW has cached it locally).

In the meantime, the SW will be updating in the background, so on the next reload they will hopefully get the latest app and everything will work fine.

There are also 6 URLs that are more problematic, because the case-sensitive versions are different as well:

- `api/core/inject-0`
- `api/core/Inject-1`
- `api/core/Injector-0`
- `api/core/INJECTOR-1`
- `api/router/Routes-0`
- `api/router/ROUTES-1`

So, if a user visits the API list page with an old SW and clicks on `Inject`, the URL will change to `api/core/Inject-1` (and the app will request `Inject-1.json` and fail as before). However, even if they reload the specific page after the SW has been updated, they will still be on `api/core/Inject-1` (with the `-1` suffix), which the new app version does not know how to handle.

##
As a work-around for this last issue, we could teach the new `DocumentService` to also be able to translate these old disambiguated URLs.

For the more general first problem (that affects all API URLs, but should go away after a reload), I am not sure if it is worth devising a work-around or accepting that it will affect users visiting the API pages before having their SW updated and would need to reload to fix it.

petebacondarwin commented:
Is it possible to know whether the SW is in the process of updating? If so, then perhaps we could do something with 404s such that we hold off on displaying the 404 page while the SW completes and then retry the request with the new SW when registered?

gkalpak commented:
> Is it possible to know whether the SW is in the process of updating?

Not really 😞
(We could hack something together, but it would require manually triggering update checks on every request, which would be expensive and compete for network resources with the actual request. Especially on initial navigation.)

> If so, then perhaps we could do something with 404s such that we hold off on displaying the 404 page while the SW completes and then retry the request with the new SW when registered?

This is more about making the right request than making the request with the new SW. So, we would need to do a full page reload, so that we get the updated `DocumentService`, which would know how to translate URLs to file names.

And whatever we did would only solve future similar issues - it would not do anything for the changes in this PR, since the currently cached apps would not know how to do this trick :smiley:

petebacondarwin commented:
> And whatever we did would only solve future similar issues - it would not do anything for the changes in this PR, since the currently cached apps would not know how to do this trick 😃

My thought was that we could land such a change before this one - leave it a week or so - then land this PR. Then there is a chance that a bunch of people will have received the preceding update which would make this change no problem for them.

gkalpak commented:
> My thought was that we could land such a change before this one - leave it a week or so - then land this PR. Then there is a chance that a bunch of people will have received the preceding update which would make this change no problem for them.

We could do this with a tweaked `DocumentService` that does the 404 handling (for example, first requests `FormBuilder.json` and if that fails requests `f_ormb_uilder.json`). Then after some time land this PR, which actually changes the file names.

WDYT?

IgorMinar commented on aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:
```diff
@@ -14,55 +14,30 @@
+ * by an underscore.
+ */
+function encodeToLowercase(str) {
+  return str.replace(/[A-Z]/g, char => char.toLowerCase() + '_');
```
`_` is a legal character in a JavaScript symbol name so we need to watch out for it.

Please consider either double-encoding it `Foo_bar` -> `_foo__bar` or use `-` char for encoding instead (`Foo_bar` -> `-foo_bar`).

petebacondarwin replied:
Since doc paths can also be non-API based then we also have to be careful with other seperater characters like `-`.
I did consider all this, and came to the conclusion that it would be rare for there to be a collision. E.g. you would need one page to be `MyPage` and the other to be `m_yp_age`. To look at it another way, if there was a page with `_` in it such as `my_value` that would only collide with page that had originally been `mYvalue`.

This seems very unlikely.

IgorMinar replied:
Good point about non-api docs! Thanks for bringing that up.

So how about we escape _ found in the input?

I'm not worried about collisions as much as a potential need to perform reverse transformation in the future.

If we escape the _ char in the input to be __, the transformation will have this property:

t(X) == Y && t'(Y) == X 

Where X is the original input, Y is the encoded output, t is the transformation and t' is a transformation that reverses t.

And that property is very nice to have to prevent future issues, especially considering how easy it is to obtain it.

petebacondarwin replied:
Yes, let's go with the bijection :-) I'll update the PR tomorrow.

IgorMinar replied:
"bijection"! I just learned something new today. Thanks for sharing 😊

IgorMinar commented on aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.spec.js:
```diff
@@ -9,12 +9,12 @@ describe('disambiguateDocPaths processor', () => {
-      {docType: 'test-doc', id: 'unique-doc', path: 'unique/doc', outputPath: 'unique/doc.json'},
-      {docType: 'test-doc', id: 'other-doc', path: 'other/doc', outputPath: 'other/doc.json'},
-      {docType: 'test-doc', id: 'other-DOC', path: 'other/DOC', outputPath: 'other/DOC.json'},
+      { docType: 'test-doc', id: 'test-doc', path: 'test/doc', outputPath: 'test/doc.json' },
```
I assume that all these diffs are due to prettier now incrementally reformatting the code base.

If that's the case, could we do a one full reformatting pass and commit that so that we remove the noise from normal PRs?

petebacondarwin replied:
We don't have a tool ensured convention for formatting in our AIO tooling. We should agree one and then run a full format pass as you suggest. Would you be happy with `Prettier` or `Clang`?

IgorMinar replied:
Ah, so these diffs are just to locally configured code formatter and not something enforced by the CI?

In that case don't worry about this PR, but in the future we might want to avoid locally configured formatters until we have a CI enforcement.

I know that there is an ongoing discussion about code formatting in the monorepo. Maybe aio should be the pilot for rolling that out. (And lastly we should avoid formating the aio/ folder with clang format because the output is not great, prettier is the way to go.)

IgorMinar reviewed (approved):
This looks very good! Can you please address the issue with `_` being a potentially problematic character? thanks @petebacondarwin

IgorMinar commented:
I see that @gkalpak raised some additional concerns around SW caching. Those seem to be valid and will need to be addressed as well.

petebacondarwin commented:
> I see that @gkalpak raised some additional concerns around SW caching. Those seem to be valid and will need to be addressed as well.

Yeah - I'll create a new PR that prepares the SW for this change.

mary-poppins commented:
You can preview 51b7c93 at https://pr42414-51b7c93.ngbuilds.io/.

mary-poppins commented:
You can preview 9ea2604 at https://pr42414-9ea2604.ngbuilds.io/.

mary-poppins commented:
You can preview 6626751 at https://pr42414-6626751.ngbuilds.io/.

mary-poppins commented:
You can preview 6140186 at https://pr42414-6140186.ngbuilds.io/.

mary-poppins commented:
You can preview 48b69ff at https://pr42414-48b69ff.ngbuilds.io/.

jelbourn reviewed (approved):
LGTM

Reviewed-for: size-tracking

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: size-tracking

mary-poppins commented:
You can preview 1661870 at https://pr42414-1661870.ngbuilds.io/.
